### PR TITLE
Fix for issue #233

### PR DESCRIPTION
**Pull request by Nebiroth**
Electron version defaults to port 3000 on startup, but will attempt to change ports if the first port is busy.

### CODE REVIEW

File: src/application/common/endpoint.ts [hunk 1]
```diff
@@ -24,7 +27,21 @@ export class Endpoint {
     }
 
     protected get host() {
-        return location.host || "127.0.0.1:3000"
+
+
+        if (location.host === "") {
+            this.fs.readFile('port', "utf8", function (err: string, data: any) {
```
Comment (akosyakov) [inline src/application/common/endpoint.ts]: `port` just should be injected, don't need to share it is over file

File: src/application/node/application.ts [hunk 2]
```diff
@@ -23,26 +23,54 @@ export interface BackendApplicationContribution {
...
+    createFile(portNumber: number) {
+
+        this.fs.writeFile('port', portNumber.toString(), function (err: string) {
+            if (err) {
+                return console.error(err);
+            }
+        });
+    }
+
+    start(portNumber: number): Promise<void> {
```
Comment (akosyakov) [inline src/application/node/application.ts]: `port` is good name to me, e.g. it is used in most APIs, only if it is undefined we should auto pick one, if not we should fail

File: src/application/node/application.ts [hunk 3]
```diff
@@ -23,26 +23,54 @@ export interface BackendApplicationContribution {
...
     constructor(
         @inject(ContributionProvider) @named(BackendApplicationContribution)
         protected readonly contributionsProvider: ContributionProvider<BackendApplicationContribution>,
         @inject(ILogger) protected readonly logger: ILogger
     ) { }
 
-    start(port: number = 3000): Promise<void> {
+    createFile(portNumber: number) {
+
+        this.fs.writeFile('port', portNumber.toString(), function (err: string) {
```
Comment (akosyakov) [inline src/application/node/application.ts]: we should use injection instead

File: examples/electron/package.json [hunk 4]
```diff
@@ -31,8 +31,9 @@
     "build:frontend": "webpack --target electron && cp src/frontend/index.html lib/frontend",
     "build:localdeps": "ldm run build && ldm sync --original-sources",
     "lint": "tslint -c ../../config/tslint/tslint.json --project ./tsconfig.json",
-    "start": "electron ./lib/frontend/electron-main.js | bunyan",
+    "start": "bunyan | electron ./lib/frontend/electron-main.js ",
     "start:debug": "electron ./lib/frontend/electron-main.js --loglevel=debug | bunyan",
+    "start:portTest": "npm run start -- 3005",
```
Comment (akosyakov) [inline examples/electron/package.json]: it should not appear in the final PR

File: src/application/browser/application.ts [hunk 5]
```diff
@@ -26,6 +31,7 @@ export interface FrontendApplicationContribution {
 export class FrontendApplication {
 
     protected _shell: ApplicationShell | undefined;
+    public static isBrowser: boolean = true;
```
Comment (akosyakov) [inline src/application/browser/application.ts]: why do you need it?

File: src/application/common/endpoint.ts (hunk 1 above)
Comment (akosyakov) [inline src/application/common/endpoint.ts]: also you cannot use node APIs in common modules, please review code organization: https://github.com/theia-ide/theia/wiki

File: src/application/node/application.ts (hunk 2 above)
Comment (akosyakov) [inline src/application/node/application.ts]: if the port is undefined then it should be zero in that case node net should assign random port

File: src/application/browser/application.ts [hunk 6]
```diff
@@ -17,7 +17,12 @@ import { Widget } from "./widgets";
 export const FrontendApplicationContribution = Symbol("FrontendApplicationContribution");
 export interface FrontendApplicationContribution {
     /**
-     * Callback
+     *CallbackThe value of the $TRAVIS_OS_NAME variable is
```
Comment (akosyakov) [inline src/application/browser/application.ts]: Why is it here?

File: src/application/common/endpoint.ts (hunk 1 above)
Comment (hexa00) [inline src/application/common/endpoint.ts]: @akosyakov The idea of the file is to use it in webpack also... or whatever the production proxy is... maybe we could use injection inside the app but still write it in one place for external clients?

Comment (akosyakov) [inline src/application/common/endpoint.ts]: With the latest merged PR port and host are provided to `generator-theia` via the yeoman configuration file: https://github.com/theia-ide/theia/blob/master/examples/browser/.yo-rc.json

We could generate the webpack config and the backend code are using the same port and host for the web with `generator-theia`.

For electron i think it should be possible to open the window with a properly configured location (that contains host) after the express has been started and the port is known.

Comment (akosyakov) [inline src/application/common/endpoint.ts]: for electron I meant something like:
```ts
require("../backend/main").then(app => {
    mainWindow = new electron.BrowserWindow({ width: 1024, height: 728 });
    mainWindow.webContents.openDevTools();
    mainWindow.loadURL(`http://localhost:${app.port}`);
    // or as a parameter to a file URI
    // mainWindow.loadURL(`file://${path.join(__dirname, '../../lib/index.html')}?port=${app.port}`);
    mainWindow.on('closed', function () {
        mainWindow = undefined;
    });
});
```

where `app.port` is implemented via `server.address().port`

**Comment by akosyakov** [comment]
Plese, use rebase instead of merge to catch up with the master and don't forget to sign off all you commit.

Maybe you better open a new PR with just a minimal change for electron first?

**Comment by Nebiroth** [comment]
Last commit was not intentional. I will close this PR now and make a small electron change with an OS-chosen port.

**Pull request closed without merging**